### DISCUSSION TITLE
Add specific Entity Hiding to Plugin API

### DISCUSSION THREAD

**Discussion by Marstead** [runelite/runelite#14977] [Ideas]
**How it works right now:** Right now, plugin authors don't have access to hide or unhide individual entities (ore nodes, trees, NPCs, enemies, etc). Entity Hiding is an "All or Nothing" thing through the native Entity Hider plugin, but currently plugins cannot control the hide/unhide state of individual entities.

**Proposed Change:** Add the ability for plugin authors to hide or unhide individual entities (interactable skilling objects & NPCs) via the API.

**Potential Problems:** After discussing this with Adam in the Discord, part of the issue with adding this feature is that NPC entities have inconsistent internal IDs; server restarts and game updates can change them. So if a plugin hides a specific NPC, say a specific goblin or cow, that hide state might not survive the next restart or update if the selected NPC ID is changed. Or worse, it might get misapplied to the wrong, now-shifted ID.

**Proposed Solution:** There are other existing use cases in RuneLite that only work on a per-session basis. For example, the respawn timer feature only applies to NPC tracked during a single session; any tracked respawn timer data is automatically cleared & reset when RuneLite is closed or the server restarts. If there are concerns around targeted entity hiding in custom plugins because of inconsistent IDs, reset all hide/unhide data whenever a session ends (RuneLite closes) or the server restarts.

**Example Use Case:** It is already useful to hide some entities, for example hiding players at Wintertodt or hiding attackers at Nechryaels/Dust Devils to make it easier to stack them after drawing their aggro. However, the Entity Hider is all or nothing, so the player has to constantly toggle on and off when moving between activities. Having more control over hiding, even if it is only on a per-session basis, could be convenient for certain activities--for example, a custom plugin that only hides attacking NPCs typically associated with auto-retaliate barraging such as Dust Devils. In my particular use case, I'd like to be able to temporarily hide entities during a session as part of a restricted account concept.

**Comment by Enriath** [comment]
The entity hider is all or nothing to prevent abuse. Under [Jagex's statement](https://secure.runescape.com/m=news/another-message-about-unofficial-clients?oldschool=1), outright allowing any particular NPC to be hidden would be directly against multiple different sections of said statement; The plugin as it is has at the moment been deemed fine (which is still surprising to me)
